### DISCUSSION TITLE
JavaScript Intellisense Issues

### DISCUSSION THREAD

**Discussion by TurtIeSocks** [pmndrs/zustand#1951] [Bug report] [answered]
Follow up from https://github.com/pmndrs/zustand/issues/262#issuecomment-1646103830

I have an older project I wrote in JS and is just a little too tedious to port entirely to TS, and I noticed that the intellisense doesn't seem to work when `set` and `get` (or only `set`) arguments are used when creating the store.

For example:
Creating this store:
![Screenshot 2023-07-21 at 2 38 01 PM](https://github.com/pmndrs/zustand/assets/58572875/730825b1-5414-4427-8fa4-f4f6312c4c60)
Results in this intellisense:
![Screenshot 2023-07-21 at 2 37 49 PM](https://github.com/pmndrs/zustand/assets/58572875/64f39851-8763-4875-bf4d-59156b8ec88b)

But creating this store:
![Screenshot 2023-07-21 at 2 38 12 PM](https://github.com/pmndrs/zustand/assets/58572875/10ed1d13-eeec-4688-aa1c-71ef1d383b8d)
Results in no intellisense:
![Screenshot 2023-07-21 at 2 38 27 PM](https://github.com/pmndrs/zustand/assets/58572875/8199c41f-105f-43c4-8725-1ad2edb620c8)

The results are the same for each scenario when wrapping the store with the `persist` middleware.

Version:
`"zustand": "^4.3.9"`

Zustand is imported like so (if it's helpful to know)
```js
import { create } from 'zustand'
import { persist } from 'zustand/middleware'
```

Any suggestions? 

Thank you for this wonderful lib! 🙏

**Comment by dai-shi** [comment]
Thanks for opening up a discussion.
I hope someone can dig into it.

Just out of curiosity, does changing `create(...)` to `create()(...)` change anything?

**Reply by TurtIeSocks** [reply]
It doesn't, unfortunately. It also caused the loss of intellisense in the scenario where set and get args aren't provided.

**Reply by prenaissance** [reply]
```ts
export type StateCreator<
  T,
  Mis extends [StoreMutatorIdentifier, unknown][] = [],
  Mos extends [StoreMutatorIdentifier, unknown][] = [],
  U = T,
> = ((
  setState: Get<Mutate<StoreApi<T>, Mis>, 'setState', never>,
  getState: Get<Mutate<StoreApi<T>, Mis>, 'getState', never>,
  store: Mutate<StoreApi<T>, Mis>
) => U) & { $$storeMutators?: Mos }
```
I think it boils down to the circular dependency between the initialization parameters and the `T` generic, making inference impossible. I tried rolling back Typescript's language server and the same result was replicated.

**Comment by prenaissance** [comment, marked as answer]
[Here](https://codesandbox.io/s/elastic-borg-ncj9jw?file=/src/store.js)'s quite a verbose solution using JSDoc.
![image](https://github.com/pmndrs/zustand/assets/95138356/6a035985-8866-494e-a4ed-1534031bb953)

**Reply by TurtIeSocks** [reply]
Thanks for this :) about the same as TS anyway (unless I've been using it wrong) and definitely better than nothing.
